core/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala

```diff
@@ -57,24 +61,75 @@ abstract class OptimizeTableCommandBase extends RunnableCommand with DeltaComman
 case class OptimizeTableCommand(
     path: Option[String],
     tableId: Option[TableIdentifier],
-    partitionPredicate: Option[String])
+    partitionPredicate: Option[String])(val zOrderBy: Option[Seq[UnresolvedAttribute]])
   extends OptimizeTableCommandBase with LeafRunnableCommand {
 
+  override val otherCopyArgs: Seq[AnyRef] = zOrderBy :: Nil
+
   override def run(sparkSession: SparkSession): Seq[Row] = {
     val deltaLog = getDeltaLog(sparkSession, path, tableId, "OPTIMIZE")
 
+    val partitionColumns = deltaLog.snapshot.metadata.partitionColumns
     // Parse the predicate expression into Catalyst expression and verify only simple filters
     // on partition columns are present
     val partitionPredicates = partitionPredicate.map(predicate => {
       val predicates = parsePredicates(sparkSession, predicate)
       verifyPartitionPredicates(
         sparkSession,
-        deltaLog.snapshot.metadata.partitionColumns,
+        partitionColumns,
         predicates)
       predicates
     }).getOrElse(Seq.empty)
 
-    new OptimizeExecutor(sparkSession, deltaLog, partitionPredicates)
+    val zOrderByColumns = if (zOrderBy.isEmpty) {
+      Seq.empty
+    } else {
+      // Create a `DataFrame` for the table schema without partition columns
```

Comment:
can this code be moved into a separate method? seems like a big distraction from the main flow in `run`